File: frontend/src/app/services/template/template.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Page } from '../../models/page';
import { TemplateDTO } from '../../models/template.dto';
import { TemplateCommandDTO } from '../../models/templateCommand.dto';
import { CommandDTO } from '../../models/command.dto';

/**
 * Servicio encargado de la gestión de las plantillas
 */
@Injectable({
  providedIn: 'root'
})
export class TemplateService {
  private http = inject(HttpClient);
  private apiUrl = environment.apiUrl;
  private readonly baseUrl = this.apiUrl + '/template/';

  /**
   * 
   * @param literal texto utilizado para buscar por nombre o descripción
   * @param page página a mostrar
   * @param size tamaño por página
   * @returns Plantillas paginadas
   */
  searchTemplates(literal: string = '', page: number = 0, size: number = 10): Observable<Page<TemplateDTO>> {
    let url = `${this.baseUrl}search?page=${page}&size=${size}`;
    if (literal.trim()) {
      url += `&literal=${encodeURIComponent(literal.trim())}`;
    }
    return this.http.get<Page<TemplateDTO>>(url);
  }

  /**
   * 
   * @param template Plantilla a añadir
   * @returns 
   */
  addTemplate(template: TemplateDTO) : Observable<void> {
    const headers = new HttpHeaders({
      'Content-Type': 'application/json'
    });
    return this.http.post<void>(this.baseUrl + "register", template, { headers })
  }

  /**
   * 
   * @param template Plantilla a modificar
   * @returns 
   */
  updateTemplate(template: TemplateDTO) : Observable<void> {
    const headers = new HttpHeaders({
      'Content-Type': 'application/json'
    });
    return this.http.post<void>(this.baseUrl + "update", template, { headers })
  }

  /**
   * 
   * @param id Identidicador de plantilla a eliminar
   * @returns 
   */
  deleteTemplateById(id:number): Observable<void> {
    return this.http.delete<void>(this.baseUrl + "delete/" + id)
  }

  /**
   * 
   * @param templateId Identificador de plantilla
   * @returns Comandos asignados a un plantilla
   */
  getAssignedCommands(templateId: number): Observable<TemplateCommandDTO[]> {
    return this.http.get<TemplateCommandDTO[]>(`${this.baseUrl}${templateId}/commands`);
  }
  
  /**
   * 
   * @param templateId Identificador de plantilla
   * @returns Comandos disponibles para asignar a la plantilla
   */
  getAvailableCommands(templateId: number): Observable<CommandDTO[]> {
    return this.http.get<CommandDTO[]>(`${this.baseUrl}${templateId}/available-commands`);
  }
  
  /**
   * 
   * @param templateId Identificador de plantilla
   * @param commands Comandos a asignar a la plantilla
   * @returns 
   */
  assignCommandsToTemplate(templateId: number, commands: TemplateCommandDTO[]): Observable<void> {
    const headers = new HttpHeaders({
      'Content-Type': 'application/json'
    });
    return this.http.post<void>(`${this.baseUrl}${templateId}/assign-commands`, commands, { headers });
  }
}

File: backend/src/main/java/com/gca/service/TemplateService.java
package com.gca.service;

import com.gca.dto.CommandDTO;
import com.gca.dto.TemplateCommandDTO;
import com.gca.dto.TemplateDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Interfaz que define las operaciones del servicio de plantillas.
 */
public interface TemplateService {

    /**
     * Crea una nueva plantilla.
     *
     * @param templateDTO los datos de la plantilla a crear
     * @return el ID de la plantilla creada
     */
    Long createTemplate(TemplateDTO templateDTO);

    /**
     * Actualiza una plantilla existente.
     *
     * @param templateDTO los datos de la plantilla a actualizar
     * @return el ID de la plantilla actualizada
     */
    Long updateTemplate(TemplateDTO templateDTO);

    /**
     * Elimina una plantilla por su ID.
     *
     * @param id el ID de la plantilla a eliminar
     */
    void deleteTemplate(Long id);

    /**
     * Busca una plantilla por literal.
     * @param literal el literal a buscar en las plantillas
     * @param pageable los parámetros de paginación
     * @return una página de plantillas que coinciden con el literal
     */
    Page<TemplateDTO> searchTemplate(String literal, Pageable pageable);

    /**
     * Obtiene los comandos asignados a una plantilla.
     * @param templateId el ID de la plantilla
     * @return una lista de comandos asignados a la plantilla
     */
    List<TemplateCommandDTO> getAssignedCommands(Long templateId);

    /**
     * Obtiene los comandos disponibles para asignar a una plantilla.
     * @param templateId el ID de la plantilla
     * @return una lista de comandos disponibles
     */
    List<CommandDTO> getAvailableCommands(Long templateId);

    /**
     * Asigna comandos a una plantilla.
     * @param templateId el ID de la plantilla
     * @param comandos la lista de comandos a asignar
     */
    void assignCommandsToTemplate(Long templateId, List<TemplateCommandDTO> comandos);
}
